ast/asdl_rs.py

```diff
@@ -346,6 +342,14 @@ def sum_subtype_struct(self, sum_type_info, t, rust_name, depth):
             depth,
         )
 
+        self.emit(f"impl{generics_applied} Ranged for {payload_name}{generics_applied} {{", depth)
```

Comment:
```suggestion
        ranged_applied = self.apply_generics(t.name, "<TextRange>")
        self.emit(f"impl Ranged for {payload_name}{ranged_applied} {{", depth)
```
And changing here too to implement it only for `U=TextTrange`

Reply:
That makes sense. I'm not sure if these changes make sense for RustPython. Maybe it's something that only Ruff wants. 

I can see how it can be convenient to "remove" the source information from the nodes or replace them with row/column (but the way this is implemented today is probably relatively slow). The challenge I'm facing with ruff is that:

* We need the start and end positions for every node. The alternative would be to wrap every node with `Attributed,` but that feels complicated
* We want to use the `expr.as_constant()` method and pass the `Constant` around while retaining the byte offset positions. Implementing this with `Attributed` would require implementing all the `as`, and `into` methods for `Attributed<ExprKind>` too. We can do this but, at least for ruff, the source is so essential that it makes sense to always pass it.

Reply:
If you'd like to add range to every nodes, adding a feature like `fully-ranged` will be possible. With the feature, every nodes even without attributes can have extra `custom` field. Adding `custom:U` for previously attributed types and adding `#[cfg(feature = "fully-ranged")] custom: U` for previously un-attributed types. That looks totally okay to RustPython. 

Because `Attributed<T>` is implementing `Deref<Target=T>`, `expr.as_consonant()` already works.

Reply:
> Because Attributed<T> is implementing Deref<Target=T>, expr.as_consonant() already works.

The problem is that this returns a `Constant` but we need a `Constant` with the source information.

Reply:
You are right. Because Ruff don't require `Fold` trait, maybe generating another module by feature makes more sense. 
- `generic` -> `ranged`   which enabled by feature `ranged`
- `located` -> `located`   which enabled by feature `location` (depending of ranged)
- `rustpython_ast::fully_ranged` <- adding new one by feature. which enabled by feature `fully_ranged`

all the names are instantly named. They can be named better.

User must enable one of ranged or fully_ranged to use any kind of ast.
Still it will share most of infrastructure.

Parser is the problem. We can't make a pair of parser. Giving up additive features concept for parser?

Reply:
I think adding non-additive feature is better than fork for that reason. Unless any user tries to use both rustpython parser and ruff at the same time, it will not break.
That sounds very unlikely to happen and if really needed, fork can be executed at that time.

Reply:
Making compatible parser for with and without the feature will be not a big deal. lalrpop can always try to pass range but ast can ignore it. Then it will be optimized out to no-op.

Reply:
Do I understand it correctly that you recommend adding a new feature that wraps nodes with `Attributed` regardless if they have the `attributes` specified in the asdl? 

We would then sprinkle the AST, parser, and all other RustPython crates with `cfg![]` to either access the node directly or restructure the `Attributed`? 

I need to think about this. I think I would still prefer having the location on the node itself from an ergonomics perspective (it's such vital information when writing a formatter or linter) but I can see that it may not make sense for RustPython.

Reply:
> Do I understand it correctly that you recommend adding a new feature that wraps nodes with Attributed regardless if they have the attributes specified in the asdl?

I think removing Attributed will be a lot easier to do that. Adding a field by feature is a lot more controllable than adding a new type wrapper by feature.

And then only the new fields need to be handled with cfg and everything others the same.

Reply:
> Does Ruff need range everywhere as cache? Or is there also missing information?

We need it for every node so that we can associate comments with nodes. But we don't need it for tokens (e.g. operators). 

> At a glance, most of non-attributed nodes are Expr and expr seems to hold the nodes' location eventually except for operator as text itself (like +)

Some exceptions are `comprehension`, `arguments`, `with_item` and `match_case`. 

It's also somewhat more convenient to pass the node directly than always having to write `Attributed<Node>` but that's neglectable.

Reply:
with attributes,

previous:
```rust
#[derive(Clone, Debug, PartialEq, is_macro::Is)]
pub enum ExprKind<U = ()> {
    #[is(name = "bool_op_expr")]
    BoolOp(ExprBoolOp<U>),
   ...
}
pub type Expr<U = ()> = Attributed<ExprKind<U>, U>;

#[derive(Clone, Debug, PartialEq)]
pub struct ExprBoolOp<U = ()> {
    pub op: Boolop,
    pub values: Vec<Expr<U>>,
}
```

new:

```rust
#[derive(Clone, Debug, PartialEq, is_macro::Is)]
pub enum Kind<U = ()> {
    #[is(name = "bool_op_expr")]
    BoolOp(ExprBoolOp<U>),
   ...
}

#[derive(Clone, Debug, PartialEq)]
pub struct ExprBoolOp<U = ()> {
    pub custom: U,
    pub op: Boolop,
    pub values: Vec<Expr<U>>,
}
```

Without attributes

previous:
```rust
#[derive(Clone, Debug, PartialEq)]
pub struct Comprehension<U = ()> {
    pub target: Expr<U>,
    pub iter: Expr<U>,
    pub ifs: Vec<Expr<U>>,
    pub is_async: bool,
}
```

new:
```rust
#[derive(Clone, Debug, PartialEq)]
pub struct Comprehension<U = ()> {
    #[cfg(feature = "fully-attributed")] pub custom: U,
    pub target: Expr<U>,
    pub iter: Expr<U>,
    pub ifs: Vec<Expr<U>>,
    pub is_async: bool,
}
```

Does it make sense?

Reply:
Ah, I got it. Comprehension used by expr is okay but used by ListComp is not attributed. Isn't this python caveat? Maybe worth to raise issue to python too.
I think RustPython also can have location information for those items to generate better messages.
What about other nodes like `ModModule`? If you don't need location there, whitelisting them looks more reasonable.

Reply:
> What about other nodes like ModModule? If you don't need location there, whitelisting them looks more reasonable.

Oh yeah, I missed that one. We want locations for modules to, for example we must know that the module range includes the body and all type_ignores in case of  `Module(stmt* body, type_ignore* type_ignores)` 

I guess an alternative could be (I would need to look at the lexing) to manually implement `range` for module as:

```
fn range(&self) {
	TextRange::new(self.body().start(), self.type_ignores.end())
}
```

But that would require that `TypeIgnore` has location information... so we're turning in cycles :)

Reply:
Module node is rarely created, so it is not a big deal actually. (though adding to type_ignore looks enough)

How about others?
- expr_context: I don't believe this has location info
- boolop, unaryop, cmpop: Expr has information, but it can be not enough
- operator: if operator symbol doesn't need to be located, it is known in expr
- comprehension: obviously missing
- arguments, withitem, matchcase: looking calculatable? if it is not used by hot path, adding a helper methods for those nodes will make sense.

Add location to comprehension looks good. Typeignore is fine.

I guess adding location to operator, boolop, unaryop and cmpop will affect RustPython. 
If any of those are still need to be added, I prefer to add a feature.

If possible, arguments, withitem, matchcase also are preferred to be avoided.